### DISCUSSION TITLE
How to recalculate bantime/endofban fields in database, after revising formula?

### DISCUSSION THREAD

**Discussion by MrPeteH** [fail2ban/fail2ban#3476] [Q&A] [answered]
I have a live system with thousands of legitimate bans. (I've been around a long time and attract a lot of attacking bots :( )

I just finished properly tuning my incremental ban formula... and because my old formula was far off (some bans are for 200+ years!) I'd like to do something that seems it should be both Simple and Easily Accomplished....

  _Recalculate the bantime and endban time for existing bans_

Yet I don't see that as a feature. Am I missing something? Would this be as simple (hah) as some kind of sqlite3 query, or does it also need to modify the firewall/ipset/???

**Comment by sebres** [comment, marked as answer]
Persistently recalculate?.. Why? 
If they get unbanned but attacking again later, they will be banned using your new formula / parameters (once you define it and `reload` configuration). If they don't come back, it is OK to forget them at some point (consider a defined purge time and after delay 3 times longer than a last known bantime by IP), since they are not recidive or even dynamic IPs possibly.

But OK, sure you can do that,... but at the moment using SQL statements only. There is no such feature as reban by fail2ban at the moment, however you can possibly try to use a rescan solution from https://github.com/fail2ban/fail2ban/issues/2909#issuecomment-1039267423 (just I'm unsure it'd work for active bans, they should be probably unbanned firstly).

Regarding the statements:
This one would show all active bans in fail2ban database, you can also remove `where endofban > datetime('now', 'localtime')` to show all known "bad" IPs (not yet purged by auto-cleanup).
```bash
sqlite3 -header -column 'file:/var/lib/fail2ban/fail2ban.sqlite3?mode=ro' "select datetime(timeofban, 'unixepoch', 'localtime') as startofban, datetime(timeofban + bantime, 'unixepoch', 'localtime') as endofban, ip, jail, bantime, bancount, data from bips where endofban > datetime('now', 'localtime') order by jail, endofban desc"
```

Something like that would update them using SQL formula (here a default one):
```
bantime=3600; # 1h
banfactor=1; 
maxtime=3024000; #5 weeks
sqlite3 'file:/var/lib/fail2ban/fail2ban.sqlite3' "update bips set bantime = min($maxtime, $bantime * (1<<(iif(bancount<20, bancount, 20))) * $banfactor) where endofban > datetime('now', 'localtime')"
```
You can also adjust the formula and change where-criteria (e. g. to update single jail only).
This would surely expect a restart of fail2ban after the update.

**Reply by MrPeteH** [reply]
Apologies for the delay, @sebres ! RL stuff...

First, why: I accidentally set up a bantime formula that put a number of IP's a few HUNDRED years into the future! So no, not likely to do a reban anytime soon LOL...

As a pragmatic matter, what I did this time as a workaround:
* Selected all bans whose bantime is larger than anything I want to see -- this validated what I was about to do
* Did a sqlite ... "update bips set bantime = 0..." for those -- caused a one-time unban
* The next time they're seen, they automagically get a correct bantime
* (Yes, this brings a small risk. It presumes that my jails are effective at catching bad guys doing what bad guys do. In my context, I'm willing to make that assumption :-D )

I like your solution even better, as it puts the actual formula in... and allows me to play with the formula if/as needed. This is better than many other methods, in that it retains the bancount and other historical information.
